### ISSUE TITLE
A11y: Product cards not sufficiently structured

### ISSUE THREAD

**Issue by blacksheep--** [shopware/platform#7100]
*PHP Version:* 8.2   | *Shopware Version:* 6.6.9.0  | *Affected area / extension:* Platform(Default) 

 ----

*_Actual behaviour:_*

Product cards are structured in a way that makes it hard for screen reader users to comprehend.

1. The cards main information (i.e. the product title) is not marked as a heading or list element. Screen reader users can not use list navigation features to navigate between product cards.
2. The description and price are in divs separated from the title, so the card will not be read out as a contained block of information together with the title. It is hard for a screen reader user to comprehend this should be a contained block of information (title, description, price, link to detail page).
3. The image is only of decorative use, so the current alt-tag (which is the product name) leads to the the product name being read out two times in a row. 

*_Expected behaviour:_*

1. The card's title should be marked as heading or the card should be a list element to allow screen reader users to navigate between them.
2. Description and price should be marked as semantic text blocks (ie. `<p></p>`) or at least semantically interconnected with the title (e.g. aria-describedby, aria-labelledby)
3. The alt-tag of the image (which is of decorative use only) should be set to alt="" to avoid reading out the product name twice in a row

Resources:
- [Inclusive Components - Cards](https://inclusive-components.design/cards/)

*_How to reproduce:_*

Enable `ACCESSIBILITY_TWEAKS` Feature-Flag

1. navigate to a page showing product cards (e.g. category or wishlist)

For better referencing use NVDA Screen Reader.

**Comment by shopwareBot** [comment]
[public] Shopware automated response: This issue is linked to the internal issue: **NEXT-40853**



[created from NEXT-40853, comment 541866]

**Comment by shopwareBot** [comment]
[public] 

 > Priority changed to: **High**

 Thank you for your report. Based on the information in this report, we have decided to investigate this issue in more detail and are planning to provide a fix for this issue. Although we are always eager to solve such problems as soon as we can, we can't give you a more specific time frame right now. We appreciate your patience and thank you for your understanding.



[created from NEXT-40853, comment 541991]

**Comment by BrocksiNet** [comment]
related: #5237

**Comment by tobiasberge** [comment]
Hi @blacksheep-- thank you for helping us to improve accessibility. 👍 

We have tried to address your feedback in this [PR](https://github.com/shopware/shopware/pull/7414). 

Because shopware is highly extendable and the rendered content can also be dynamic we often run into limitations that prevent us from applying some best practices right away. 

I wanted to share my thoughts with your points:

**1. cards title:** totally agree, this should have a headline like most cards out there. We introduced a setting for this because the product listing is part of the CMS. The content is dynamic and there can be other existing headline structure on the same page already. So we did not want to hard-code h2 for everyone. But h2 is the recommended default.

**2. description and price:** I am not sure here about aria-describedby, aria-labelledby and linking it to the products title. If I understand the documentation correctly they are intended to be used on interactive elements like `a` or `button` and it is not recommended to use them on static texts. `Don't use aria-label or aria-labelledby on any other non-interactive content such as p, legend, li, or ul, because it is ignored.` https://www.w3.org/TR/using-aria/#practical-support-aria-label-aria-labelledby-and-aria-describedby 

Even if I add a aria-labelledby to the description text (pointing to the headline) it will read the product title again and produce a "group" that I have to enter and exit again (At least VoicerOver does this). For me this created even more "noise" and makes it more complicated. 

I was also thinking about creating a landmark region/section with aria-label "prices for productName" to wrap the pricing. But then documentation says to [use landmarks sparingly](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/region_role#accessibility_concerns). And I assume that landmarks in each product box would be too much.

For now, we have added `visually-hidden` labels to better describe the prices (especially when there are multiple because of discounts etc.) But we are happy to discuss if you have other ideas.

**3. image without alt:** We have not removed the alt text for now. I am not sure here if the image is really purely decorative since it is the main product cover image. And not a "mood image" for a blog post preview of something similar. It is also not always the case that the alt text equals exactly the product title. The shop owner can enter individual alt titles in the media manager that are then used. Also, the screen reader announces the image as "image". Additionally, I thought it is considered good practice to have the experience/user-flow as close as possible in the screen reader as for a screen user. But also here, feel free to share your thoughts if you have more insights. This is not set in stone.

**Comment by blacksheep--** [comment]
Hey @tobiasberge thank your for your detailed answer. What I think is important is that cards are read out as concise as possible. When a screen reader user navigates between cards, it should be easily recognisable (without any further action) that:

- this is a product
- the product's name/title
- the product's (short) description
- the product's price
- (optional) the product has variants
- more information is available on the product's detail page

Imagine someone describing the content of the webpage to you, will you be able to grasp the necessary information? What would you like to hear, what don't you need. The flow/experience will always be different for someone solely relying on heared information - the medium is a different one.
